denote.el

```diff
@@ -3933,6 +3937,143 @@ To be assigned to `markdown-follow-link-functions'."
+minor-mode `denote-fontify-links-mode' appropriately. See the
+documentation of that function for more details.
+
+When the value is `fontify', the link is fontified using
+font-lock. The appropriate keyword must be added to `font-lock'.
+Enabling the minor-mode `denote-fontify-links-mode' does this.
+
+The value `fontify-and-fold', is the same as `fontify' but also
+uses `org-fold' (if available) to fold the links.
+"
+  :type '(choice
+          (const :tag "As button" button)
+          (const :tag "Fontified" fontify)
+          (const :tag "Fontified and folded" fontify-and-fold))
+  :package-version '(denote . "0.6.0")
+  :group 'denote)
+
+(defvar denote-link-keymap
+  (let ((map (make-sparse-keymap)))
+    (define-key map [mouse-2] #'denote-link-open-at-mouse)
+    (define-key map [mouse-3] #'denote-link-open-at-mouse)
+    (define-key map [follow-link] #'mouse-face)
+    map)
+  "Keymap used in fontified denote links")
+
+(defun denote-link-open-at-mouse (ev)
+  "Open denote link after mouse click."
+  (interactive "e")
+  (mouse-set-point ev)
+  (when-let ((id (get-text-property (point) 'denote-link-id)))
+    (funcall denote-link-button-action (denote-get-path-by-id id))))
+
+(defun denote-fontify-links (&optional limit)
+  "Fontify denote links.
+
+Implementation based on `org-activate-links--text-properties'"
+  (catch :exit
+    (when-let (type (denote-filetype-heuristics (buffer-file-name)))
+      (while (re-search-forward
+              (denote--link-in-context-regexp type)
+              limit t)
+        (save-match-data  ;; to return the matches to font-lock
+          (let* ((start (match-beginning 0))
+	         (end (match-end 0))
+	         (visible-start (match-beginning 2))
+	         (visible-end (match-end 2))
+                 (id (match-string-no-properties 1))
+                 (link (concat "file:" (denote-get-path-by-id id))))
+	    (unless (let ((face (get-text-property
+			         (max (1- start) (point-min)) 'face)))
+		      (if (consp face)
+                          (memq 'font-lock-comment-face face)
+		        (eq 'font-lock-comment-face face)))
+	      (when (and (featurep 'org-fold)
+                         (eq denote-link-display-style 'fontify-and-fold))
+                (org-fold-region start end nil 'org-link)
+                (org-fold-region start end nil 'org-link-description)
+                ;; We are folding the whole emphasized text with 'org-link
+                ;; first.  It makes everything invisible (or whatever
+                ;; the user wants).
+                (org-fold-region start end t 'org-link)
+                ;; The visible part of the text is folded using
+                ;; 'org-link-description, which is forcing this part of
+                ;; the text to be visible.
+                (org-fold-region visible-start visible-end t
+                                 'org-link-description)
+                (org-rear-nonsticky-at visible-start)
```

Comment:
A lot of this code is specific to Org which is a problem for us in practice because `denote.el` does not `(require 'org)`.  Of course, we still have the blocking issue with `org-fold` that I mentioned above.

Reply:
I can see that denote already has several calls to `(org-*) functions, `denote-link-ol-*` and `denote-date-prompt`. In any case, the code is protected with a `featurep` check for `org-fold`.